### PR TITLE
fix jit buffer reuse when freed

### LINKED ISSUE #1799: Stable Diffusion Memory Usage on M1 8GB is 11.62 GB
Clip:
```
***    0 r_386_32_12352_4_4                arg   1 sz [386]              [32]         OPs   2441M/   0.00G  mem  4.27 GB tm  10836.92us/    10.84ms (  225.27 GFLOPS,    0.02 GB/s)
```

First step:
```
***    1 r_5_5                             arg   1 sz []                 []           OPs      0M/   0.00G  mem  3.77 GB tm      7.29us/     0.01ms (    0.00 GFLOPS,    0.00 GB/s)
```

Second step:
```
***    1 E_                                arg   2 sz []                 []           OPs      0M/   0.00G  mem  3.77 GB tm      6.92us/     0.02ms (    0.00 GFLOPS,    0.00 GB/s)
```

Third step:
```
***    1 E_                                arg   2 sz []                 []           OPs      0M/   0.00G  mem 11.62 GB tm      6.88us/     0.02ms (    0.00 GFLOPS,    0.00 GB/s)
```

### PR DESCRIPTION
**Pull request by nimlgen**
Closes #1799

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/jit.py                           26              15
tinygrad/ops.py                            3               0
tinygrad/runtime/lib.py                    2               0
------------------------------------------------------------
total                                     31              15
------------------------------------------------------------
lines added in the tinygrad folder: 16
```

File: tinygrad/jit.py [hunk 1]
```diff
@@ -64,37 +65,41 @@ def __call__(self, *args, **kwargs) -> Any:
...
-    def alloc_rawbuf(self): return self.buftype(self.size, self.dtype, **({'device':self.device} if self.device is not None else dict()))
+    def alloc_rawbuf(self): return self.buftype(self.size, self.dtype, **({'device':self._device} if self._device is not None else dict()))
 
   def __init__(self):
     self.cache: Optional[List[Tuple[Callable, List[Any], Dict[Any,Any]]]] = None
     self.placeholders: Dict[RawBuffer, _CacheCollector._Placeholder] = {} # Rawbuffers are replaced with placeholders to allow freeing of the real buffer while collecting cache.
     self.last_buftype: Dict[Tuple[int,...], int] = {} # Last index of the cached entry where a buffer with the shape (shape is a key) is used as input to the prog.
     self.last_placeholder_index: Dict[_CacheCollector._Placeholder, int] = {} # Last index where the placeholder is used as output. This allows tracking when we need to stick to the original buffer if it is still alive.
+    self.freed_placeholders: Dict[Tuple[int,...], List[_CacheCollector._Placeholder]] = defaultdict(list)
+    self.cirular_signatures: Set[Any] = set()
```
Comment (geohot) [inline tinygrad/jit.py:78]: typo

Comment (nimlgen) [reply tinygrad/jit.py]: done

**Comment by geohot** [comment]
Thanks! I don't fully understand this code, but trust this fixes it.

**Pull request merged**